File: leetcode/swap-nodes-in-pairs/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}
	if k <= 1 {
		return head
	}
	nodesLen := 0
	for p := head; p != nil; p = p.Next {
		nodesLen++
	} 
    dummy := &ListNode{
        Next:head,
    }
	dummy.Next = head	
	pre := dummy
    tail, next := head, head
	for; nodesLen >= k; nodesLen-=k {
		for i := 0; i < k; i++ {
			temp := next.Next
			next.Next = pre.Next
			pre.Next = next
			next = temp
			tail.Next = next
		}
		pre = tail
		tail = next
	}
	return dummy.Next
}

File: leetcode-golang/Swap Nodes in Pairs/Swap Nodes in Pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func swapPairs(head *ListNode) *ListNode {
    list := &ListNode{
		Next: head
	}
	for prev, node := list, list.Next; node != nil ; node = node.Next {
		if node.Next != nil {
			swapNode(prev, node, node.Next)
			prev = node
		}
	}
	return list.Next
}

//每两个相邻节点交换
func swapNode(prev, node, next *ListNode) {
	prev.Next = next
	node.Next = next.Next
	next.Next = node
}